File: credit_expert_bot.py
#!/usr/bin/env python3
"""
Credit Expert Bot - AI-powered assistant for debt relief services
Saves all users and messages to MySQL database
"""

import asyncio
import os
import logging
import re
from typing import Optional
import pymysql
from telethon import TelegramClient, events
from telethon.tl.types import User
from dotenv import load_dotenv
from integrations.openai_client import OpenAIClient


# Emoji removal pattern
EMOJI_PATTERN = re.compile(
    "["
    "\U0001f600-\U0001f64f"  # emoticons
    "\U0001f300-\U0001f5ff"  # symbols & pictographs
    "\U0001f680-\U0001f6ff"  # transport & map symbols
    "\U0001f1e0-\U0001f1ff"  # flags
    "\U00002702-\U000027b0"  # dingbats
    "\U000024c2-\U0001f251"  # enclosed characters
    "\U0001f900-\U0001f9ff"  # supplemental symbols
    "\U0001fa00-\U0001fa6f"  # chess symbols
    "\U0001fa70-\U0001faff"  # symbols extended-A
    "\U00002600-\U000026ff"  # misc symbols
    "\U00002700-\U000027bf"  # dingbats
    "]+",
    flags=re.UNICODE,
)

# Name validation pattern (Russian/English names)
NAME_PATTERN = re.compile(r"^[А-Яа-яЁёA-Za-z][А-Яа-яЁёA-Za-z\-\s]{1,30}$")


def strip_emoji(text: str) -> str:
    """Remove emoji from text"""
    return EMOJI_PATTERN.sub("", text).strip()


def is_valid_name(name: str) -> bool:
    """Check if text looks like a valid name"""
    name = name.strip()
    # Too short or too long
    if len(name) < 2 or len(name) > 30:
        return False
    # Contains only letters, spaces, hyphens (for double names)
    if not NAME_PATTERN.match(name):
        return False
    # Probably not a name if contains common non-name words
    non_names = [
        "привет",
        "здравствуй",
        "добрый",
        "вечер",
        "день",
        "утро",
        "да",
        "нет",
        "ок",
        "окей",
        "хорошо",
        "ладно",
        "понял",
        "спасибо",
        "пожалуйста",
        "извини",
        "простите",
    ]
    if name.lower() in non_names:
        return False
    return True


# Load environment
load_dotenv()

# Logging setup
logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
logger = logging.getLogger("Credit_Expert_Bot")

# Telegram credentials
API_ID = int(os.getenv("TELEGRAM_API_ID"))
API_HASH = os.getenv("TELEGRAM_API_HASH")
BOT_TOKEN = os.getenv("CREDIT_EXPERT_BOT_TOKEN")  # Assuming a new token env var
SESSION_FILE = "credit_expert_bot"

# MySQL config
MYSQL_CONFIG = {
    "host": os.getenv("MYSQL_HOST", "localhost"),
    "port": int(os.getenv("MYSQL_PORT", 3306)),
    "database": os.getenv("MYSQL_DATABASE", "pythorust_tg"),
    "user": os.getenv("MYSQL_USER", "pythorust_tg"),
    "password": os.getenv("MYSQL_PASSWORD"),
    "charset": "utf8mb4",
    "cursorclass": pymysql.cursors.DictCursor,
}

# System prompt based on user instruction
CREDIT_EXPERT_SYSTEM_PROMPT = """Ты — кредитный эксперт Дарья из ФЦБ (Федеральный Центр Банкротства).
ТВОЯ ЦЕЛЬ: Получить номер телефона клиента для консультации.

СТИЛЬ ОБЩЕНИЯ:
- Профессиональный и деловой тон
- Без эмодзи и восклицательных знаков
- Краткие, чёткие ответы (2-4 предложения)
- Уверенный эксперт, а не подруга

ИНСТРУКЦИЯ ПО ДИАЛОГУ:

ШАГ 1: ЗНАКОМСТВО
"Добрый день. Я Дарья, кредитный эксперт ФЦБ. Как могу к вам обращаться?"

ШАГ 2: КВАЛИФИКАЦИЯ
После получения имени спроси: "[Имя], расскажите коротко вашу ситуацию — какой примерно долг и есть ли просрочки?"

Уточняющие вопросы (по одному за раз):
- Коллекторы уже звонят?
- Долги только в МФО или банки тоже?
- Есть ли имущество, которое хотите сохранить?

ШАГ 3: ПЕРЕХОД К ЗВОНКУ
"[Имя], чтобы дать конкретные рекомендации по вашей ситуации, предлагаю созвониться на 10 минут. Консультация бесплатная. Какой номер для связи?"

ОТРАБОТКА ВОЗРАЖЕНИЙ:
- "Расскажите сначала": "Каждый случай индивидуален. На звонке за 10 минут разберём ваш — это эффективнее переписки."
- "Сколько стоит?": "Зависит от суммы долга и ситуации. Консультация бесплатная — расскажу варианты и стоимость."
- "Подумаю": "Хорошо. Учтите, что пени и проценты продолжают расти. Готова ответить, когда решите."
- "Можно в переписке?": "В переписке сложно учесть все нюансы. 10 минут разговора заменят час переписки. Давайте попробуем?"

СТРОГИЕ ПРАВИЛА:
- Никаких эмодзи
- Никаких "подружеских" выражений
- Цель — телефон для консультации
- Один вопрос за раз
- Не затягивай переписку — веди к звонку
"""


class MySQLLogger:
    """Handles all MySQL database operations"""

    def __init__(self):
        self.conn = None

    def connect(self):
        """Establish MySQL connection"""
        self.conn = pymysql.connect(**MYSQL_CONFIG)
        logger.info("Connected to MySQL")

    def close(self):
        """Close MySQL connection"""
        if self.conn:
            self.conn.close()

    def ensure_connection(self):
        """Ensure connection is alive"""
        if not self.conn or not self.conn.open:
            self.connect()
        try:
            self.conn.ping(reconnect=True)
        except Exception:
            self.connect()

    def save_user(self, user: User) -> None:
        """Save or update user in database"""
        self.ensure_connection()

        query = """
            INSERT INTO bot_users (id, username, first_name, last_name, language_code, is_premium, is_bot)
            VALUES (%s, %s, %s, %s, %s, %s, %s)
            ON DUPLICATE KEY UPDATE
                username = VALUES(username),
                first_name = VALUES(first_name),
                last_name = VALUES(last_name),
                language_code = VALUES(language_code),
                is_premium = VALUES(is_premium),
                last_seen_at = CURRENT_TIMESTAMP
        """

        with self.conn.cursor() as cursor:
            cursor.execute(
                query,
                (
                    user.id,
                    user.username,
                    user.first_name,
                    user.last_name,
                    getattr(user, "lang_code", None),
                    getattr(user, "premium", False),
                    user.bot if hasattr(user, "bot") else False,
                ),
            )
        self.conn.commit()
        logger.info(f"Saved user: {user.id} ({user.first_name})")

    def save_message(
        self,
        user_id: int,
        message_id: int,
        text: str,
        direction: str,
        bot_name: str = "Credit_Expert_Bot",
        reply_to: Optional[int] = None,
    ) -> None:
        """Save message to database"""
        self.ensure_connection()

        query = """
            INSERT INTO bot_messages
            (telegram_message_id, user_id, bot_name, direction, message_text, reply_to_message_id)
            VALUES (%s, %s, %s, %s, %s, %s)
        """

        with self.conn.cursor() as cursor:
            cursor.execute(query, (message_id, user_id, bot_name, direction, text, reply_to))
        self.conn.commit()
        logger.info(f"Saved {direction} message for user {user_id}")

    def get_session(self, user_id: int, bot_name: str = "Credit_Expert_Bot") -> Optional[dict]:
        """Get active session for user"""
        self.ensure_connection()

        query = """
            SELECT * FROM bot_sessions
            WHERE user_id = %s AND bot_name = %s AND is_active = TRUE
            ORDER BY session_start DESC LIMIT 1
        """

        with self.conn.cursor() as cursor:
            cursor.execute(query, (user_id, bot_name))
            return cursor.fetchone()

    def create_session(self, user_id: int, bot_name: str = "Credit_Expert_Bot") -> int:
        """Create new session"""
        self.ensure_connection()

        # End any existing active sessions
        end_query = """
            UPDATE bot_sessions SET is_active = FALSE, session_end = CURRENT_TIMESTAMP
            WHERE user_id = %s AND bot_name = %s AND is_active = TRUE
        """

        create_query = """
            INSERT INTO bot_sessions (user_id, bot_name, state)
            VALUES (%s, %s, 'greeting')
        """

        with self.conn.cursor() as cursor:
            cursor.execute(end_query, (user_id, bot_name))
            cursor.execute(create_query, (user_id, bot_name))
            session_id = cursor.lastrowid
        self.conn.commit()
        return session_id

    def get_conversation_history(self, user_id: int, bot_name: str = "Credit_Expert_Bot", limit: int = 20) -> list:
        """Get recent conversation history"""
        self.ensure_connection()

        query = """
            SELECT direction, message_text, created_at
            FROM bot_messages
            WHERE user_id = %s AND bot_name = %s
            ORDER BY created_at DESC
            LIMIT %s
        """

        with self.conn.cursor() as cursor:
            cursor.execute(query, (user_id, bot_name, limit))
            messages = cursor.fetchall()

        return list(reversed(messages))


class CreditExpertBot:
    """Main bot class"""

    def __init__(self):
        self.client = None
        self.db = MySQLLogger()
        self.ai = OpenAIClient()

    async def start(self):
        """Start the bot"""
        self.db.connect()

        if BOT_TOKEN:
            self.client = TelegramClient(SESSION_FILE, API_ID, API_HASH)
            await self.client.start(bot_token=BOT_TOKEN)
        else:
            logger.error("CREDIT_EXPERT_BOT_TOKEN not set!")
            # For testing purposes, we might want to allow running without token if mocked
            # but in production code it should return.
            # We'll keep it as is for now.
            return

        logger.info("Credit Expert Bot started")

        # Register handlers
        @self.client.on(events.NewMessage(pattern="/start"))
        async def start_handler(event):
            await self.handle_start(event)

        @self.client.on(events.NewMessage)
        async def message_handler(event):
            if not event.message.text.startswith("/"):
                await self.handle_message(event)

        # Run until disconnected
        await self.client.run_until_disconnected()

    async def handle_start(self, event):
        """Handle /start command"""
        user = await event.get_sender()

        # Save user to DB
        self.db.save_user(user)

        # Save incoming message
        self.db.save_message(user_id=user.id, message_id=event.message.id, text="/start", direction="incoming")

        # Create new session
        self.db.create_session(user.id)

        # Send greeting (Step 1) - professional tone, no emoji
        greeting = "Добрый день. Я Дарья, кредитный эксперт ФЦБ. Как могу к вам обращаться?"

        # Send messages
        msg1 = await event.respond(greeting)

        # Save outgoing messages
        self.db.save_message(user_id=user.id, message_id=msg1.id, text=greeting, direction="outgoing")

    async def handle_message(self, event):
        """Handle incoming messages"""
        user = await event.get_sender()
        text = event.message.text

        # Save user
        self.db.save_user(user)

        # Save incoming message
        self.db.save_message(user_id=user.id, message_id=event.message.id, text=text, direction="incoming")

        # Get or create session - ensure conversation continues
        session = self.db.get_session(user.id)
        if not session:
            self.db.create_session(user.id)

        # Get conversation history
        history = self.db.get_conversation_history(user.id)

        # Check if this might be a name response (early in conversation)
        is_name_expected = len(history) <= 2  # Just started, expecting name

        # Build messages for AI
        messages = [{"role": "system", "content": CREDIT_EXPERT_SYSTEM_PROMPT}]

        for msg in history:
            role = "assistant" if msg["direction"] == "outgoing" else "user"
            messages.append({"role": role, "content": msg["message_text"]})

        # Add current message with name validation hint if needed
        current_msg = text
        if is_name_expected and not is_valid_name(text):
            # Add hint for AI that name seems invalid
            current_msg = f"{text}\n[СИСТЕМА: Это не похоже на имя. Переспроси имя клиента вежливо.]"

        messages.append({"role": "user", "content": current_msg})

        # Get AI response
        try:
            response = await self.ai.chat_completion(messages)
            response_text = response.choices[0].message.content
            # Strip emoji from response - professional tone
            response_text = strip_emoji(response_text)
            # Clean up extra whitespace that might appear after emoji removal
            response_text = re.sub(r"\s+", " ", response_text).strip()
            response_text = re.sub(r" ([.,!?])", r"\1", response_text)
        except Exception as e:
            logger.error(f"AI error: {e}")
            response_text = "Извините, произошла техническая ошибка. Напишите ещё раз."

        # Send response
        sent_msg = await event.respond(response_text)

        # Save outgoing message
        self.db.save_message(user_id=user.id, message_id=sent_msg.id, text=response_text, direction="outgoing")

    def stop(self):
        """Stop the bot"""
        self.db.close()
        if self.client:
            self.client.disconnect()


async def main():
    bot = CreditExpertBot()
    try:
        await bot.start()
    except KeyboardInterrupt:
        bot.stop()
    except Exception as e:
        logger.error(f"Bot error: {e}")
        bot.stop()
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: src/bin/credit_expert_bot.rs
//! Credit Expert Bot (Rust)
//!
//! Telegram бот-консультант по банкротству с логированием в MySQL.
//! Переписан с Python-версии `credit_expert_bot.py`.

use std::sync::Arc;

use anyhow::{Context, Result};
use dotenvy::dotenv;
use mysql_async::{prelude::*, Pool};
use regex::Regex;
use telegram_reader::integrations::openai::ChatMessage;
use telegram_reader::integrations::OpenAIClient;
use teloxide::dispatching::UpdateFilterExt;
use teloxide::prelude::*;
use teloxide::types::{Message, User};
use tracing::{error, info};

const BOT_NAME: &str = "Credit_Expert_Bot";

const CREDIT_EXPERT_SYSTEM_PROMPT: &str = r#"Ты — кредитный эксперт Дарья из ФЦБ (Федеральный Центр Банкротства).
ТВОЯ ЦЕЛЬ: Получить номер телефона клиента для консультации.

СТИЛЬ ОБЩЕНИЯ:
- Профессиональный и деловой тон
- Без эмодзи и восклицательных знаков
- Краткие, чёткие ответы (2-4 предложения)
- Уверенный эксперт, а не подруга

ИНСТРУКЦИЯ ПО ДИАЛОГУ:

ШАГ 1: ЗНАКОМСТВО
"Добрый день. Я Дарья, кредитный эксперт ФЦБ. Как могу к вам обращаться?"

ШАГ 2: КВАЛИФИКАЦИЯ
После получения имени спроси: "[Имя], расскажите коротко вашу ситуацию — какой примерно долг и есть ли просрочки?"

Уточняющие вопросы (по одному за раз):
- Коллекторы уже звонят?
- Долги только в МФО или банки тоже?
- Есть ли имущество, которое хотите сохранить?

ШАГ 3: ПЕРЕХОД К ЗВОНКУ
"[Имя], чтобы дать конкретные рекомендации по вашей ситуации, предлагаю созвониться на 10 минут. Консультация бесплатная. Какой номер для связи?"

ОТРАБОТКА ВОЗРАЖЕНИЙ:
- "Расскажите сначала": "Каждый случай индивидуален. На звонке за 10 минут разберём ваш — это эффективнее переписки."
- "Сколько стоит?": "Зависит от суммы долга и ситуации. Консультация бесплатная — расскажу варианты и стоимость."
- "Подумаю": "Хорошо. Учтите, что пени и проценты продолжают расти. Готова ответить, когда решите."
- "Можно в переписке?": "В переписке сложно учесть все нюансы. 10 минут разговора заменят час переписки. Давайте попробуем?"

СТРОГИЕ ПРАВИЛА:
- Никаких эмодзи
- Никаких "подружеских" выражений
- Цель — телефон для консультации
- Один вопрос за раз
- Не затягивай переписку — веди к звонку
"#;

/// Remove emoji from text.
fn strip_emoji(text: &str) -> String {
    let emoji_regex = Regex::new(
        r"[\x{1f600}-\x{1f64f}\x{1f300}-\x{1f5ff}\x{1f680}-\x{1f6ff}\x{1f1e0}-\x{1f1ff}\x{2702}-\x{27b0}\x{24c2}-\x{1f251}\x{1f900}-\x{1f9ff}\x{1fa00}-\x{1fa6f}\x{1fa70}-\x{1faff}\x{2600}-\x{26ff}\x{2700}-\x{27bf}]+"
    ).unwrap();
    emoji_regex.replace_all(text, "").to_string()
}

/// Check if text looks like a valid name.
fn is_valid_name(name: &str) -> bool {
    let name = name.trim();
    if name.len() < 2 || name.len() > 30 {
        return false;
    }

    let name_regex = Regex::new(r"^[А-Яа-яЁёA-Za-z][А-Яа-яЁёA-Za-z\-\s]{1,30}$").unwrap();
    if !name_regex.is_match(name) {
        return false;
    }

    let non_names = [
        "привет",
        "здравствуй",
        "добрый",
        "вечер",
        "день",
        "утро",
        "да",
        "нет",
        "ок",
        "окей",
        "хорошо",
        "ладно",
        "понял",
        "спасибо",
        "пожалуйста",
        "извини",
        "простите",
    ];

    !non_names.contains(&name.to_lowercase().as_str())
}

/// MySQL logger for bot data.
#[derive(Clone)]
struct MySqlLogger {
    pool: Pool,
}

impl MySqlLogger {
    fn new(pool: Pool) -> Self {
        Self { pool }
    }

    async fn save_user(&self, user: &User) -> Result<()> {
        let mut conn = self.pool.get_conn().await?;

        conn.exec_drop(
            r#"
            INSERT INTO bot_users (id, username, first_name, last_name, language_code, is_premium, is_bot)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            ON DUPLICATE KEY UPDATE
                username = VALUES(username),
                first_name = VALUES(first_name),
                last_name = VALUES(last_name),
                language_code = VALUES(language_code),
                is_premium = VALUES(is_premium),
                last_seen_at = CURRENT_TIMESTAMP
            "#,
            (
                user.id.0 as i64,
                &user.username,
                &user.first_name,
                &user.last_name,
                user.language_code.as_deref(),
                user.is_premium,
                false,
            ),
        )
        .await?;

        info!(user_id = user.id.0, name = %user.first_name, "Saved user");
        Ok(())
    }

    async fn save_message(
        &self,
        user_id: i64,
        message_id: i32,
        text: &str,
        direction: &str,
        reply_to: Option<i32>,
    ) -> Result<()> {
        let mut conn = self.pool.get_conn().await?;

        conn.exec_drop(
            r#"
            INSERT INTO bot_messages
            (telegram_message_id, user_id, bot_name, direction, message_text, reply_to_message_id)
            VALUES (?, ?, ?, ?, ?, ?)
            "#,
            (message_id, user_id, BOT_NAME, direction, text, reply_to),
        )
        .await?;

        info!(user_id = user_id, direction = direction, "Saved message");
        Ok(())
    }

    async fn get_or_create_session(&self, user_id: i64) -> Result<i64> {
        let mut conn = self.pool.get_conn().await?;

        // Check for existing active session
        let existing: Option<i64> = conn
            .exec_first(
                r#"
                SELECT id FROM bot_sessions
                WHERE user_id = ? AND bot_name = ? AND is_active = TRUE
                ORDER BY session_start DESC LIMIT 1
                "#,
                (user_id, BOT_NAME),
            )
            .await?;

        if let Some(session_id) = existing {
            return Ok(session_id);
        }

        // End any existing sessions and create new one
        conn.exec_drop(
            r#"
            UPDATE bot_sessions SET is_active = FALSE, session_end = CURRENT_TIMESTAMP
            WHERE user_id = ? AND bot_name = ? AND is_active = TRUE
            "#,
            (user_id, BOT_NAME),
        )
        .await?;

        conn.exec_drop(
            r#"
            INSERT INTO bot_sessions (user_id, bot_name, state)
            VALUES (?, ?, 'greeting')
            "#,
            (user_id, BOT_NAME),
        )
        .await?;

        let session_id: i64 = conn
            .exec_first("SELECT LAST_INSERT_ID()", ())
            .await?
            .unwrap_or(0);

        Ok(session_id)
    }

    async fn get_conversation_history(
        &self,
        user_id: i64,
        limit: u32,
    ) -> Result<Vec<(String, String)>> {
        let mut conn = self.pool.get_conn().await?;

        let rows: Vec<(String, String)> = conn
            .exec(
                r#"
                SELECT direction, message_text
                FROM bot_messages
                WHERE user_id = ? AND bot_name = ?
                ORDER BY created_at DESC
                LIMIT ?
                "#,
                (user_id, BOT_NAME, limit),
            )
            .await?;

        // Reverse to get chronological order
        Ok(rows.into_iter().rev().collect())
    }
}

/// Application state shared across handlers.
#[derive(Clone)]
struct AppState {
    db: Arc<MySqlLogger>,
    ai: OpenAIClient,
}

async fn handle_start(bot: Bot, msg: Message, state: AppState) -> Result<()> {
    let user = msg.from().context("No user in message")?;
    let user_id = user.id.0 as i64;

    // Save user
    state.db.save_user(user).await?;

    // Save incoming message
    state
        .db
        .save_message(user_id, msg.id.0, "/start", "incoming", None)
        .await?;

    // Create new session
    state.db.get_or_create_session(user_id).await?;

    // Send greeting
    let greeting = "Добрый день. Я Дарья, кредитный эксперт ФЦБ. Как могу к вам обращаться?";
    let sent = bot.send_message(msg.chat.id, greeting).await?;

    // Save outgoing message
    state
        .db
        .save_message(user_id, sent.id.0, greeting, "outgoing", None)
        .await?;

    Ok(())
}

async fn handle_message(bot: Bot, msg: Message, state: AppState) -> Result<()> {
    let text = match msg.text() {
        Some(t) if !t.starts_with('/') => t,
        _ => return Ok(()),
    };

    let user = msg.from().context("No user in message")?;
    let user_id = user.id.0 as i64;

    // Save user
    state.db.save_user(user).await?;

    // Save incoming message
    state
        .db
        .save_message(user_id, msg.id.0, text, "incoming", None)
        .await?;

    // Ensure session exists
    state.db.get_or_create_session(user_id).await?;

    // Get conversation history
    let history = state.db.get_conversation_history(user_id, 20).await?;

    // Check if name is expected (early in conversation)
    let is_name_expected = history.len() <= 2;

    // Build messages for AI
    let mut messages = vec![ChatMessage {
        role: "system".to_string(),
        content: Some(CREDIT_EXPERT_SYSTEM_PROMPT.to_string()),
    }];

    for (direction, msg_text) in &history {
        let role = if direction == "outgoing" {
            "assistant"
        } else {
            "user"
        };
        messages.push(ChatMessage {
            role: role.to_string(),
            content: Some(msg_text.clone()),
        });
    }

    // Add current message with name validation hint if needed
    let current_msg = if is_name_expected && !is_valid_name(text) {
        format!(
            "{}\n[СИСТЕМА: Это не похоже на имя. Переспроси имя клиента вежливо.]",
            text
        )
    } else {
        text.to_string()
    };

    messages.push(ChatMessage {
        role: "user".to_string(),
        content: Some(current_msg),
    });

    // Get AI response
    let response_text = match state
        .ai
        .chat_completion(messages, "gpt-4o-mini", 0.7, 500)
        .await
    {
        Ok(text) => {
            // Strip emoji and clean up whitespace
            let cleaned = strip_emoji(&text);
            let cleaned = Regex::new(r"\s+")
                .unwrap()
                .replace_all(&cleaned, " ")
                .trim()
                .to_string();
            Regex::new(r" ([.,!?])")
                .unwrap()
                .replace_all(&cleaned, "$1")
                .to_string()
        }
        Err(e) => {
            error!(error = %e, "AI error");
            "Извините, произошла техническая ошибка. Напишите ещё раз.".to_string()
        }
    };

    // Send response
    let sent = bot.send_message(msg.chat.id, &response_text).await?;

    // Save outgoing message
    state
        .db
        .save_message(user_id, sent.id.0, &response_text, "outgoing", None)
        .await?;

    Ok(())
}

fn build_mysql_url() -> String {
    let host = std::env::var("MYSQL_HOST").unwrap_or_else(|_| "localhost".to_string());
    let port = std::env::var("MYSQL_PORT").unwrap_or_else(|_| "3306".to_string());
    let database = std::env::var("MYSQL_DATABASE").unwrap_or_else(|_| "pythorust_tg".to_string());
    let user = std::env::var("MYSQL_USER").unwrap_or_else(|_| "pythorust_tg".to_string());
    let password = std::env::var("MYSQL_PASSWORD").unwrap_or_default();

    format!(
        "mysql://{}:{}@{}:{}/{}",
        user, password, host, port, database
    )
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    tracing_subscriber::fmt::init();

    let token = std::env::var("CREDIT_EXPERT_BOT_TOKEN")
        .context("CREDIT_EXPERT_BOT_TOKEN not set in environment (.env)")?;

    // Initialize MySQL pool
    let pool = Pool::new(build_mysql_url().as_str());

    // Initialize OpenAI client
    let ai = OpenAIClient::from_env()?;

    let state = AppState {
        db: Arc::new(MySqlLogger::new(pool)),
        ai,
    };

    info!("Starting Credit Expert Bot...");

    let bot = Bot::new(token);

    // Simple dispatcher that handles both /start and regular messages
    Dispatcher::builder(
        bot,
        Update::filter_message().endpoint(move |bot: Bot, msg: Message| {
            let state = state.clone();
            async move {
                if msg.text() == Some("/start") {
                    handle_start(bot, msg, state).await
                } else {
                    handle_message(bot, msg, state).await
                }
            }
        }),
    )
    .enable_ctrlc_handler()
    .build()
    .dispatch()
    .await;

    Ok(())
}
